File: Sources/YToolsCore/RelativePanelPlacement.swift
/// A screen-independent top-left position measured within a display's usable area.
public struct RelativePanelPlacement: Equatable, Sendable {
    public let horizontalFraction: Double
    public let topFraction: Double
    public let sourceVisibleWidth: Double
    public let sourceVisibleHeight: Double

    public init?(
        horizontalFraction: Double,
        topFraction: Double,
        sourceVisibleWidth: Double,
        sourceVisibleHeight: Double
    ) {
        guard horizontalFraction.isFinite,
              topFraction.isFinite,
              (0...1).contains(horizontalFraction),
              (0...1).contains(topFraction),
              sourceVisibleWidth.isFinite,
              sourceVisibleHeight.isFinite,
              sourceVisibleWidth > 0,
              sourceVisibleHeight > 0 else { return nil }
        self.horizontalFraction = horizontalFraction
        self.topFraction = topFraction
        self.sourceVisibleWidth = sourceVisibleWidth
        self.sourceVisibleHeight = sourceVisibleHeight
    }

    public init?(
        left: Double,
        top: Double,
        visibleOriginX: Double,
        visibleOriginY: Double,
        visibleWidth: Double,
        visibleHeight: Double
    ) {
        guard left.isFinite,
              top.isFinite,
              visibleOriginX.isFinite,
              visibleOriginY.isFinite,
              visibleWidth.isFinite,
              visibleHeight.isFinite,
              visibleWidth > 0,
              visibleHeight > 0 else { return nil }
        let maximumY = visibleOriginY + visibleHeight
        horizontalFraction = min(
            max((left - visibleOriginX) / visibleWidth, 0),
            1
        )
        topFraction = min(max((maximumY - top) / visibleHeight, 0), 1)
        sourceVisibleWidth = visibleWidth
        sourceVisibleHeight = visibleHeight
    }

    public func resolvedLeft(visibleOriginX: Double, visibleWidth: Double) -> Double {
        visibleOriginX + visibleWidth * horizontalFraction
    }

    public func resolvedTop(visibleOriginY: Double, visibleHeight: Double) -> Double {
        visibleOriginY + visibleHeight * (1 - topFraction)
    }
}

File: src/YTools.Windows/Core/RelativePanelPlacement.cs
namespace YTools.Core;

/// <summary>
/// A screen-independent top-left position measured within a display's usable
/// area. Fractions use a top-down axis to mirror the macOS port.
/// </summary>
public readonly record struct RelativePanelPlacement(
    double HorizontalFraction,
    double TopFraction,
    double SourceVisibleWidth,
    double SourceVisibleHeight)
{
    public static RelativePanelPlacement? Create(
        double horizontalFraction,
        double topFraction,
        double sourceVisibleWidth,
        double sourceVisibleHeight)
    {
        if (!double.IsFinite(horizontalFraction)
            || !double.IsFinite(topFraction)
            || horizontalFraction is < 0 or > 1
            || topFraction is < 0 or > 1
            || !double.IsFinite(sourceVisibleWidth)
            || !double.IsFinite(sourceVisibleHeight)
            || sourceVisibleWidth <= 0
            || sourceVisibleHeight <= 0)
        {
            return null;
        }

        return new RelativePanelPlacement(
            horizontalFraction,
            topFraction,
            sourceVisibleWidth,
            sourceVisibleHeight);
    }

    public static RelativePanelPlacement? Create(
        double left,
        double top,
        double visibleOriginX,
        double visibleOriginY,
        double visibleWidth,
        double visibleHeight)
    {
        if (!double.IsFinite(left)
            || !double.IsFinite(top)
            || !double.IsFinite(visibleOriginX)
            || !double.IsFinite(visibleOriginY)
            || !double.IsFinite(visibleWidth)
            || !double.IsFinite(visibleHeight)
            || visibleWidth <= 0
            || visibleHeight <= 0)
        {
            return null;
        }

        var maximumY = visibleOriginY + visibleHeight;
        var horizontalFraction = Math.Clamp((left - visibleOriginX) / visibleWidth, 0, 1);
        var topFraction = Math.Clamp((maximumY - top) / visibleHeight, 0, 1);
        return new RelativePanelPlacement(
            horizontalFraction,
            topFraction,
            visibleWidth,
            visibleHeight);
    }

    public double ResolvedLeft(double visibleOriginX, double visibleWidth)
    {
        return visibleOriginX + visibleWidth * HorizontalFraction;
    }

    public double ResolvedTop(double visibleOriginY, double visibleHeight)
    {
        return visibleOriginY + visibleHeight * (1 - TopFraction);
    }
}
